torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -648,6 +651,12 @@ bool ProcessGroupNCCL::WorkNCCL::wait(std::chrono::milliseconds timeout) {
       static_cast<int>(devices_.size())); // worldSize
   synchronizeInternal(timeout);
   // Always return true, because abort API is not implemented.
+  if (distDebugLevel_ >= DebugLevel::Detail) {
+    auto numel = getTensorsNumel(*outputs_);
+    auto hashValue = hashTensors(*outputs_);
```

Comment:
I chatted with @fduwjj, there are thoughts to remove `outputs_` from work objects, to avoid holding output buffer's memory for longer than needed. 

I am wondering if the hashing, if needed, can be called at the user side of ProcessGroupNCCL than within ProcessGroupNCCL.

By "user side" I mean:
- called by the user at python level; or
- called within `allgather_object_list` at python level.

There, there are explicit handle to the output tensor; if any follow-up operation is performed on the tensor, extension of the tensor's lifecycle is easily understandable.

Here, `work->outputs_` is an implicit reference. It would be great if we don't have to keep it, which may lead to some surprising effects (like build-up of peak memory).